llama.cpp

```diff
@@ -11754,7 +11779,7 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
         ggml_backend_tensor_set(lctx.inp_pos, batch.pos, 0, n_tokens*ggml_element_size(lctx.inp_pos));
     }
 
-    if (hparams.causal_attn || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE) {
+    if (!cparams.embeddings || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE) {
```

Comment:
Hmm. Some outputs should still be skipped when embedding for some of the pooling types, no?

This will cause uninitialized `lctx.inp_out_ids` when embedding with non-Bert models with pooling types other than NONE.

This condition was there originally for how BERT managed input skipping.

https://github.com/ggerganov/llama.cpp/blob/f8ec8877b75774fc6c47559d529dac423877bcad/llama.cpp#L8534

Since `batch.logits` is likely correctly set when using pooled embeddings (at least, how you wrote them seems correct), then should this condition instead always be true?

And if that is done, then `inp_cls` would be redundant, since the correct rows would already be the only thing left.

Feel free to make objections.

Reply:
Yeah, that makes sense. I'm guessing we want to avoid putting a `pooling_type == LLAMA_POOLING_TYPE_NONE` in every single other model? In that case, I guess we have to actually require all logits be set when getting non-NONE embeddings from non-Bert models. The downside is that it results in a needless `get_rows` on all the outputs.

In fact, it seems like `batch.logits` isn't really used when `pooling_type` is not `NONE`, since we use all the outputs and the results are stored in `embd_seq_out`. Or actually, all that's currently required is that at least one logit is requested so you go down the right branch when we check if `lctx.n_outputs == 0` in `llama_decode_internal`. It seems like in this case we might want to officially ignore `batch.logits` and give priority to `cparams.embeddings`.

Reply:
I think the simpler way to fix this in the meantime is to make `n_outputs == n_tokens_all` in `llama_decode_internal` for all non-NONE pooling types when `cparams.embeddings` is true. It could then re-use the same logic as `logits_all`.

But I think the CLS and LAST pooling types could eventually skip computing the embeddings they don't need.

Reply:
Ok, I think this should do it. Basically bypass logits when doing non-NONE embeddings. Note that I'm using `hparams.causal_attn` to decide if we're in a BERT model or not in `llama_set_inputs`.